# 1.1 - General Info of LTCM
    -Securities Traded: LTCM exploited market mispricings and sought . To achieve that, they engaged in convergence and relative value trading strategies. They sold liquidity or in other words, they became the counterparty to financial institutions that had specific needs (French insurance companies or Japanese banks) - dislocation in financial markets. There is a wide range of securities they traded. Primarily, they traded government bond swaps spreads (exchanging fixed interest with variable interest or vice versa). Also, they traded fixed rate residential mortgages, volatility, relative value in yield curve and in equities.
    
    -Trading Freq: The goal of LTCM was to make money in the long horizon. Especially for their convergence trades, they took long positions on one side and hedged that position with a short on the other side. Some of the trades liquidated as quickly as 6 months.
    
    -Skewness of Trades: They targetted small gains from many trades using leverage to exploit arbitrage. However, they are susceptible to extreme market events
    
    -Forecasting: Their models found mispricing in the market and the principals of LTCM found the raionale of the mispricing. Specifically, they avoided outright long positions. The preference of trades were in majority little to no default ones. However, to exploit relative value strategy, for instance they included positions of long of a corporation's debt and short of its equity.

# 1.2 - Advantage of LTCM
    - Efficient Financing: Primary edge is in efficient funding and financing due to prop trading capabilities
    - Fund Size: High AUM allowed LTCM to get favorable rates
    - Collateralization: Less haircut paid
    - Long Term committment of capital from investors
    - Liquidity and Hedging (Finding liquidity and efficient financing to match while avoiding default risk)

# 1.3 - Funding Risk
    - Collateral haircut: During high stress events, collateral haircut goes up. Normally when LTCM purchases US Treasury bonds through repo agreeements which collateralized the bond itself, there is 1% haircut in the collateralized value of the bond. When there is a market disruptive event, the haircut increases to let's say 3%. Knowing that normally LTCM funds e.g. swap spread trades with collateral, when there is higher haircut, it needs to inject cash to execute the trade.
    
    - Repo maturity: Typically, repos are overnight instruments but it's possible to see repos with maturity ranging from 6 months to 1 year. LTCM entered in long-term repo contracts which is unusual. When there was credit risk, the same repos were not available to LTCM.
    
    - Equity redemption: In convergence trade, it's possible to see dramatic divergence until the two securities diverge. This may result in a margin call and investor further financing to float the trade. This is an example of Equity redemption risk for LTCM.
    
    - Loan access: During high stress times, accessing to financing can be challenging. LTCM managed to secure an unsecured loan from a syndicate of banks without a material adverse change clause (This clause makes funding available to LTCM even if they do not perform well)

# 1.4 - Quantify and Manage Liquidity Risk
    - The liquidity risk does not seem like a problem for LTCM because their portfolios are self-financing. In other words, a self-financing portfolio does not need cash injection or withdrawal to/from the portfolio. Also, LTCM and its counterparties collateralized their obligations in a two way mark-to-market process which ensures liquidity. Furthermore, by stress-testing LTCM estimated the theoretical worst-case haircuts. Thanks to these, LTCM managed to structure its financing and didn't need to liquidate its position because of adverse market move. LTCM quantified its liquidity risk by adjusting security correlation. They assumed positive correlation across all asset classes in the short-term. 

# 1.5 - Monitor Leverage 
    - LTCM did not monitor its leverage ratio explicitly given the lack of major high stress event in the market. However, they considered returning some/all investors funds back to delever. Another thing is rven though their notional amount of contract looked very high, rather than ending a contract (which is usually costly), they hedged their risk with a new contract. This pushed LTCM's total notional amount higher but did not make them necessarily highly levered.

# 1.6 - Convergence Trade & Relative Value Trade
    - Relative value trade is basically taking a relative bet against two securities rather than an absolute bet in one or both. An example from the case is stocks of Shell and Royal Dutch Petroleum. Because the merger will 100% go through, Royal Dutch Petroleum's stock price is expected to converge to Shell's stock price ultimately since they would become the same company after merger. This is the Convergence Trade part of the Relative Trade strategy. 
    
    - Shell was trading in discount at Royal Dutch, LTCM longed Shell and shorted Royal Dutch. This exposes LTCM only to the differential value risk between these two stocks. And because there is no directional bet, there is no need for hedging (Relative value trade is already immune to directional changes as long as the two securities move in the same direction)
    
    - Another example of the Convergence Trade can be from fixed income securities. There are many different interest rates such as Fed Funds Rate and SOFR (Collateralized Overnight Funds Rate). Theoretically, these are risk-free rates and they represent the same rate. However, they are not always the same in the market. If Feds Fund Rate is lower than SOFR, one can borrow from FED and long SOFR. Once the rates converge, arbitrage gives out the profit (Type II arbitrage)

    - Another example of the Convergence Trade is with futures and their underlyings. If one thinks that futures contract price is overvalued relative to the expected future price of the spot of the underlying, he can long the underlying and short the future contract. When future contract expires, the underlying and the future contract would have the same price.
    
    - One year after the case, LTCM lost most of its value due to non-converging trades it had. Convergence risk is tremendous. Note that even though ultimately these securities converge and LTCM proves itself right, securities stayed divergent for a long time and caused LTCM to blow up due to illiquidity.

# Notes on Covered Interest Parity (CIP) and Uncovered Interest Parity (UIP)
    - CIP: Covered Interest Rate Parity is a no-arbitrage relation that says the best future spot FX rate predictor is the current forward FX rate. The formula for CIP uses the forward rates and the interes rates of the base and the foreign currency to check if there is an arbitrage opportunity. However, in absence of arbitrage CIP must hold true.

    - UIP: Uncovered Interest Parity is a theory, I repeat it's a theory and therefore we can see it as a model. Unlike CIP, UIP uses the expected spot rates for the future. Ideally, the regression model(X = interest rate differential of the base and the foreign currency, y = growth in the USD against the foreign currency) should output Beta = 1 and Alpha = 0. This means that for each unit of increase in the interest rate of the foreign currency, base currency (USD) appreciates 1 unit with Alpha = 0. This boils down to no arbitrage opportunity for investor. Below, we see Beta is very close to 0 and sometimes negative or greater than 1.

    Note that if expected spot rates for the future and the forward rates are equal, than CIP and UIP point out the same equation.

# Helper Functions

In [578]:
import pandas as pd
import numpy as np
import datetime as dt
pd.set_option('display.float_format', '{:.6f}'.format)

import matplotlib.pyplot as plt
import seaborn as sns
import math

import statsmodels.api as sm
from sklearn.linear_model import LinearRegression
from scipy.stats import norm
import scipy.stats as stats

import warnings
warnings.filterwarnings("ignore")

import TA_utils as ta

from typing import Union, List

from collections import defaultdict

import re

In [579]:
# Easy to read summary stats with monthly return data!

# Always plug in a DataFrame - never pd.Series
def summary_statistics_annualized(returns, annual_factor = 12):
    """This functions returns the summary statistics for the input total/excess returns passed
    into the function"""
    
    summary_statistics = pd.DataFrame(index=returns.columns)
    summary_statistics['Mean'] = returns.mean() * annual_factor
    summary_statistics['Vol'] = returns.std() * np.sqrt(annual_factor)
    summary_statistics['Sharpe'] = (returns.mean() / returns.std()) * np.sqrt(annual_factor)
    summary_statistics['Min'] = returns.min()
    summary_statistics['Max'] = returns.max()
    summary_statistics['Skewness'] = returns.skew()
    summary_statistics['Excess Kurtosis'] = returns.kurtosis()
    summary_statistics['VaR (0.05)'] = returns.quantile(.05, axis = 0)
    summary_statistics['CVaR (0.05)'] = returns[returns <= returns.quantile(.05, axis = 0)].mean()

    wealth_index = 1000*(1+returns).cumprod()
    previous_peaks = wealth_index.cummax()
    
    drawdowns = (wealth_index - previous_peaks)/previous_peaks

    summary_statistics['Max Drawdown'] = drawdowns.min()
    summary_statistics['Peak'] = [previous_peaks[col][:drawdowns[col].idxmin()].idxmax() for col in previous_peaks.columns]
    summary_statistics['Bottom'] = drawdowns.idxmin()
    
    recovery_date = []
    for col in wealth_index.columns:
        prev_max = previous_peaks[col][:drawdowns[col].idxmin()].max()
        recovery_wealth = pd.DataFrame([wealth_index[col][drawdowns[col].idxmin():]]).T
        recovery_date.append(recovery_wealth[recovery_wealth[col] >= prev_max].index.min())
    summary_statistics['Recovery'] = recovery_date
    
    return summary_statistics

In [580]:
def calc_regression(
        y: Union[pd.DataFrame, pd.Series],
        X: Union[pd.DataFrame, pd.Series],
        intercept: bool = True,
        annual_factor: Union[None, int] = 12,
        warnings: bool = True,
        return_model: bool = False,
        calc_treynor_info_ratios: bool = True
    ):

    """
    Only plug in DataFrames for X and y whose columns are string
    """
    y = y.copy()
    X = X.copy()
    
    if 'date' in X.columns.str.lower():
        X = X.rename({'Date': 'date'}, axis=1)
        X = X.set_index('date')
    X.index.name = 'date'
    
    if warnings:
        print('"calc_regression" assumes excess returns to calculate Information and Treynor Ratios')
    if intercept:
        X = sm.add_constant(X)
    
    y_name = y.name if isinstance(y, pd.Series) else y.columns[0]
    
    try:
        model = sm.OLS(y, X, missing="drop")
    except ValueError:
        y = y.reset_index(drop=True)
        X = X.reset_index(drop=True)
        model = sm.OLS(y, X, missing="drop")
        if warnings:
            print(f'"{y_name}" Required to reset indexes to make regression work. Try passing "y" and "X" as pd.DataFrame')
    results = model.fit()
    summary = dict()

    if return_model:
        return results

    inter = results.params[0] if intercept else None
    betas = results.params[1:] if intercept else results.params

    summary["Alpha"] = inter if inter is not None else '-'
    summary["Annualized Alpha"] = inter * annual_factor if inter is not None else '-'
    summary["R-Squared"] = results.rsquared

    if isinstance(X, pd.Series):
        X = pd.DataFrame(X)

    X_assets = X.columns[1:] if intercept else X.columns
    for i, asset_name in enumerate(X_assets):
        summary[f"{asset_name} Beta"] = betas[i]

    if calc_treynor_info_ratios:
        if len([c for c in X.columns if c != 'const']) == 1:
            summary["Treynor Ratio"] = (y.mean() / betas[0])
            summary["Annualized Treynor Ratio"] = summary["Treynor Ratio"] * annual_factor
        summary["Information Ratio"] = (inter / results.resid.std()) if intercept else "-"
        summary["Annualized Information Ratio"] = summary["Information Ratio"] * np.sqrt(annual_factor) if intercept else "-"
    summary["Tracking Error"] = results.resid.std()
    summary["Annualized Tracking Error"] = results.resid.std() * np.sqrt(annual_factor)
    summary['Fitted Mean'] = results.fittedvalues.mean()
    summary['Annualized Fitted Mean'] = summary['Fitted Mean'] * annual_factor
    y_name = f"{y_name} no Intercept" if not intercept else y_name
    return pd.DataFrame(summary, index=[y_name])

In [581]:
def calc_iterative_regression(
        multiple_y: Union[pd.DataFrame, pd.Series],
        X: Union[pd.DataFrame, pd.Series],
        annual_factor: Union[None, int] = 12,
        intercept: bool = True,
        warnings: bool = True,
        calc_treynor_info_ratios: bool = True
    ):
    multiple_y = multiple_y.copy()
    X = X.copy()

    if 'date' in multiple_y.columns.str.lower():
        multiple_y = multiple_y.rename({'Date': 'date'}, axis=1)
        multiple_y = multiple_y.set_index('date')
    multiple_y.index.name = 'date'

    if 'date' in X.columns.str.lower():
        X = X.rename({'Date': 'date'}, axis=1)
        X = X.set_index('date')
    X.index.name = 'date'

    regressions = pd.DataFrame({})
    for asset in multiple_y.columns:
        y = multiple_y[[asset]]
        new_regression = calc_regression(
            y, X, annual_factor=annual_factor, intercept=intercept, warnings=warnings,
            calc_treynor_info_ratios=calc_treynor_info_ratios
        )
        regressions = pd.concat([regressions, new_regression], axis=0)
    
    return regressions

In [582]:
def format_as_percentage(value):
    return '{:.2%}'.format(value)

# 2.1 a) 2.1 b)

## Read Data

In [583]:
ltcm_ret = pd.read_excel('ltcm_exhibits_data.xlsx', sheet_name='Exhibit 2')
spy_ret = pd.read_excel('gmo_analysis_data.xlsx', sheet_name = 'returns (total)', index_col = 0)[['SPY']]
rf = pd.read_excel('gmo_analysis_data.xlsx', sheet_name = 'risk-free rate', index_col = 0)


In [584]:
# Clean the columns
ltcm_columns = [] 

for col in ltcm_ret.columns:
    my_col = ltcm_ret.iloc[1][col]
    if isinstance(my_col, float):
        continue
    if 'Performancea' in my_col or 'Performanceb' in my_col:
        my_col = my_col[:len(my_col)-1]
    ltcm_columns.append(my_col)

ltcm_columns.insert(0,'date')

ltcm_ret.columns = ltcm_columns
ltcm_ret = ltcm_ret.dropna()

ltcm_ret = ltcm_ret.set_index('date')

# Bring the first day of the month to the last day of the month
ltcm_ret.index  = pd.to_datetime(ltcm_ret.index) + pd.offsets.MonthEnd(0) 

In [585]:
# Convert the datatype of object columns to numeric. Otherwise, summary_statistics_annualized gives an error

for col in ltcm_ret.columns:
    if ltcm_ret[col].dtype == 'object':
        ltcm_ret[col] = pd.to_numeric(ltcm_ret[col], errors='coerce')
        

In [586]:
# LTCM excess return perf

spy_ret_exc = spy_ret.copy()
for col in spy_ret_exc.columns:
    spy_ret_exc[col] = spy_ret_exc[col] - rf['US3M']

ltcm_ret_exc = ltcm_ret.copy()
for col in ltcm_ret_exc.columns:
    ltcm_ret_exc[col] = ltcm_ret_exc[col] - rf['US3M']

ltcm_ret_exc = ltcm_ret_exc.rename(columns = {'Gross Monthly Performance': 'Gross Performance', 'Net Monthly Performance': 'Net Performance'})

ltcm_ret_exc = ltcm_ret_exc[ltcm_ret_exc.index.isin(rf.index)][['Gross Performance', 'Net Performance']]

perf_ltcm_exc = summary_statistics_annualized(ltcm_ret_exc, annual_factor=12)
perf_ltcm_exc

Mean      Vol   Sharpe       Min      Max  Skewness  \
Gross Performance 0.242077 0.136232 1.776946 -0.105250 0.112342 -0.287725   
Net Performance   0.155360 0.111765 1.390059 -0.105250 0.080342 -0.810239   

                   Excess Kurtosis  VaR (0.05)  CVaR (0.05)  Max Drawdown  \
Gross Performance         1.586625   -0.030445    -0.072997     -0.168949   
Net Performance           2.926921   -0.026415    -0.068664     -0.176128   

                        Peak     Bottom Recovery  
Gross Performance 1998-04-30 1998-06-30      NaT  
Net Performance   1997-12-31 1998-07-31      NaT

In [587]:
# SPY excess return perf
perf_spy_exc = summary_statistics_annualized(spy_ret_exc, annual_factor=12)
perf_spy_exc

Mean      Vol   Sharpe       Min      Max  Skewness  Excess Kurtosis  \
SPY 0.079460 0.149097 0.532939 -0.165570 0.126909 -0.558943         0.949488   

     VaR (0.05)  CVaR (0.05)  Max Drawdown       Peak     Bottom   Recovery  
SPY   -0.073525    -0.097243     -0.560012 2000-03-31 2009-02-28 2013-03-31

# 2.1 c)

##  The mean excess return and the Sharpe Ratio are dramatically high for LTCM compared to other asset classes we covered. The volatility of SPY and the gross returns are similar however VaR is much lower for LTCM

# 2.2
    a) Alpha is pretty high (0.1315) but the market Beta (SPY Beta) is quite low (~0.1371) as well as R-Squared (0.019)

    b) Closet Indexer: The term "closet indexer" refers to a type of investment fund or portfolio manager who claims to actively manage a portfolio but, in reality, closely mimics the performance of a benchmark index, such as the S&P 500, without significantly deviating from it. In other words, a closet indexer behaves more like a passive investor (index fund) while charging fees associated with active management.

    To answer the question: No, LTCM does not have a high SPY Beta and SPY does not explain the variability in LTCM returns. On the contrary, it has a very low SPY Beta and quite high Alpha which means that it is not a closet indexer. (Note that LTCM exploited long-short strategy for convergence and value trades which is the rationale behind low SPY Beta)

    c) Given the high Alpha (13.15%) beyond the market excess return, LTCM brought significant premium

In [588]:
ltcm_ret_exc_net = ltcm_ret_exc[['Net Performance']] # Net Excess Return of LTCM (Regressand)
spy_ret_ltcm_reg = spy_ret_exc[spy_ret_exc.index.isin(ltcm_ret_exc.index)] # Net Excess Return of SPY (Regressor)

ltcm_spy_perf = calc_regression(ltcm_ret_exc_net, spy_ret_ltcm_reg, calc_treynor_info_ratios=False)
ltcm_spy_perf

"calc_regression" assumes excess returns to calculate Information and Treynor Ratios
Intel MKL WARNING: Support of Intel(R) Streaming SIMD Extensions 4.2 (Intel(R) SSE4.2) enabled only processors has been deprecated. Intel oneAPI Math Kernel Library 2025.0 will require Intel(R) Advanced Vector Extensions (Intel(R) AVX) instructions.


Alpha  Annualized Alpha  R-Squared  SPY Beta  \
Net Performance 0.010961          0.131527   0.018979  0.137114   

                 Tracking Error  Annualized Tracking Error  Fitted Mean  \
Net Performance        0.031956                   0.110700     0.012947   

                 Annualized Fitted Mean  
Net Performance                0.155360

# 2.3 
    a) When square of the Market Excess return is introduced as a new regressor, Alpha increased to 15.5% per annum while r-squared is still very low (~0.025). However, the square of the market premium is dramatically inversely correlated with LTCM excess returns with a Beta of -1.9267 while the market return itself has Beta of 0.1669 (similar to the previous regression)

    b) As noted, r-squared increased from 0.190 to 0.0243 which is not a great rise. This means that addition of quadratic expression to the regression does not help me explain the variation in LTCM excess returns

    c) Because SPY^2 has negative Beta, it behaves as if it's a short market option. We can think of beta of SPY as the delta of the market option and the beta of SPY^2 as the gamma to market options 

    d) Because SPY^2 is negatively correlated with LTCM, high SPY return would cause LTCM to underperform. Therefore, LTCM is negatively exposed to market volatility

In [589]:
ltcm_ret_exc_net = ltcm_ret_exc[['Net Performance']] # Net Excess Return of LTCM (Regressand)

spy_ret_ltcm_reg = spy_ret_exc[spy_ret_exc.index.isin(ltcm_ret_exc.index)] # Net Excess Return of SPY (Regressor1)
spy_ret_ltcm_reg ['SPY^2'] = spy_ret_ltcm_reg['SPY']**2 # Square of Net Excess Return of SPY (Regressor2)

ltcm_spy_spy_2_perf = calc_regression(ltcm_ret_exc_net, spy_ret_ltcm_reg, calc_treynor_info_ratios=False)
ltcm_spy_spy_2_perf

"calc_regression" assumes excess returns to calculate Information and Treynor Ratios
Intel MKL WARNING: Support of Intel(R) Streaming SIMD Extensions 4.2 (Intel(R) SSE4.2) enabled only processors has been deprecated. Intel oneAPI Math Kernel Library 2025.0 will require Intel(R) Advanced Vector Extensions (Intel(R) AVX) instructions.


Alpha  Annualized Alpha  R-Squared  SPY Beta  SPY^2 Beta  \
Net Performance 0.012920          0.155042   0.024321  0.166878   -1.926746   

                 Tracking Error  Annualized Tracking Error  Fitted Mean  \
Net Performance        0.031869                   0.110398     0.012947   

                 Annualized Fitted Mean  
Net Performance                0.155360

# 2.4
    a) Alpha = ~10% which is still high but relatively low compared to the regression with quadratic market premium. SPY Beta is relatively high which is ~0.46. Call like factor has negative Beta -0.7821 and Put like beta is positive 1.2896. R-squared is 0.0555 which does not explain the variation on LTCM excess return much but higher than previous regression r-squared values

    b) Because Call like factor beta is negative, LTCM is short to the call like factor. Also, Put like factor has positive beta which means LTCM is long to the Put like factor.

    c) Due to higher magnitude of Put like factor (absolute value of 1.2896 is greater than absolute value of -0.7821) Put like factor moves LTCM excess return more than Call like factor

    d) LTCM shorts the Call like factor and longs the Put like factor. This again shows that LTCM is negatively exposed to market volatility. However, the negative volatility exposure comes from the combination of these two positions
    

In [590]:
k1 = .03
k2 = -.03

ltcm_ret_exc_net = ltcm_ret_exc[['Net Performance']]
spy_ret_ltcm_options_reg = spy_ret_exc[spy_ret_exc.index.isin(ltcm_ret_exc.index)] # Net Excess Return of SPY (Regressor1)

spy_ret_ltcm_options_reg['Call-like'] = np.maximum(spy_ret_exc['SPY'] - k1,0) # Call like factor (Regressor2)
spy_ret_ltcm_options_reg['Put-like'] = np.maximum(k2 - spy_ret_exc['SPY'],0) # Put like factor (Regressor3)

ltcm_options_perf = calc_regression(ltcm_ret_exc_net, spy_ret_ltcm_options_reg, calc_treynor_info_ratios=False)
ltcm_options_perf

"calc_regression" assumes excess returns to calculate Information and Treynor Ratios
Intel MKL WARNING: Support of Intel(R) Streaming SIMD Extensions 4.2 (Intel(R) SSE4.2) enabled only processors has been deprecated. Intel oneAPI Math Kernel Library 2025.0 will require Intel(R) Advanced Vector Extensions (Intel(R) AVX) instructions.


Alpha  Annualized Alpha  R-Squared  SPY Beta  \
Net Performance 0.008436          0.101231   0.055486  0.466610   

                 Call-like Beta  Put-like Beta  Tracking Error  \
Net Performance       -0.782140       1.289575        0.031356   

                 Annualized Tracking Error  Fitted Mean  \
Net Performance                   0.108620     0.012947   

                 Annualized Fitted Mean  
Net Performance                0.155360

# 3.1 a) - Question: In last year's solution t's FX rate is shifted but not t+1's FX rate which is the opposite in the formula???

    Even though excess return over Sterling and Euro Yen are negative, Swiss Franc returned positive excess return. Yen depreciated much more than Euro and Sterling


## Read Data

In [591]:
rf_rates = pd.read_excel('fx_carry_data.xlsx', sheet_name = 'risk-free rates', index_col = 0)
fx_rates = pd.read_excel('fx_carry_data.xlsx', sheet_name = 'fx rates', index_col = 0)


In [592]:
# Log conversion
rf_rates_log = rf_rates.copy()

for col in rf_rates_log.columns:
    rf_rates_log[col] = np.log(1 + rf_rates_log[col])

fx_rates_log = fx_rates.copy()

for colu in fx_rates_log.columns:
    fx_rates_log[colu] = np.log(fx_rates_log[colu])

In [593]:
rf_usd_log = rf_rates_log['USD1M']
rf_rates_log = rf_rates_log[['GBP1M', 'EUR1M', 'CHF1M', 'JPY1M']]

exc_log_ret = pd.DataFrame(columns = ['UK', 'EU', 'SZ', 'JP'], index = fx_rates.index)

for rf_col, fx_col in zip(rf_rates_log.columns, fx_rates_log.columns):
    exc_log_ret[str(fx_col)[2:]] = fx_rates_log[fx_col]- fx_rates_log[fx_col].shift(1) + rf_rates_log[rf_col].shift(1) - rf_usd_log.shift(1)
    

In [594]:
perf_exc = summary_statistics_annualized(exc_log_ret, annual_factor=12)
perf_exc

Mean      Vol    Sharpe       Min      Max  Skewness  Excess Kurtosis  \
UK -0.003502 0.086303 -0.040574 -0.094861 0.088332 -0.313949         1.389999   
EU -0.004351 0.094714 -0.045944 -0.103716 0.093510 -0.126741         1.186795   
SZ  0.004312 0.098757  0.043662 -0.118548 0.129996  0.198117         2.145865   
JP -0.017415 0.091492 -0.190342 -0.085030 0.074347 -0.180633         0.724913   

    VaR (0.05)  CVaR (0.05)  Max Drawdown       Peak     Bottom Recovery  
UK   -0.039405    -0.056968     -0.435217 2007-10-31 2019-08-31      NaT  
EU   -0.044808    -0.062149     -0.383333 2008-06-30 2020-04-30      NaT  
SZ   -0.041724    -0.056629     -0.313244 2011-07-31 2019-04-30      NaT  
JP   -0.049173    -0.062482     -0.438955 1999-11-30 2021-10-31      NaT

# 3.2 a)
    According to Uncovered Interest Parity, gains from the high interest paying FX is supposed to wash away with the depreciation of the FX against base currency (base currency is most of the time USD for us). However, we see that non of the FX has zero excess return. Even though they are quite close to zero (except Yen), currency contradict with UIP

# 3.2 b) 
    A long position in Swiss Frank gives the highest Sharpe Ratio ~0.04 even though it's super low
    
# 3.2 c)
    Yes. Sterling, Euro and Yen offer negative excess return when longed denominated by USD. Especially, Yen with the lowest excess return
    

# 3.3 a) - FX prediction - 
## Note: Positive Beta means depreciation of FX as UIP suggests -> Ideally, Beta should be 1 for 1 unit of rise in interest rate causes 1 unit of depreciation of the FX

In [595]:
fx_pred_list = []

for rf_col, fx_col in zip(rf_rates_log.columns, fx_rates_log.columns):
    y = fx_rates_log[[fx_col]] - fx_rates_log[[fx_col]].shift(1) # This is same as exc_log_ret[[fx_col]].diff()
    X = rf_usd_log - rf_rates_log[rf_col]
    X = X.to_frame()
    X = X.rename(columns={0: 'Interest Rate Differential'})
    temp_fx_pred = calc_regression(y, X)
    fx_pred_list.append(temp_fx_pred)
    
fx_pred = pd.concat(fx_pred_list)[['Annualized Alpha','Interest Rate Differential Beta', 'R-Squared']]

"calc_regression" assumes excess returns to calculate Information and Treynor Ratios
Intel MKL WARNING: Support of Intel(R) Streaming SIMD Extensions 4.2 (Intel(R) SSE4.2) enabled only processors has been deprecated. Intel oneAPI Math Kernel Library 2025.0 will require Intel(R) Advanced Vector Extensions (Intel(R) AVX) instructions.
"calc_regression" assumes excess returns to calculate Information and Treynor Ratios
Intel MKL WARNING: Support of Intel(R) Streaming SIMD Extensions 4.2 (Intel(R) SSE4.2) enabled only processors has been deprecated. Intel oneAPI Math Kernel Library 2025.0 will require Intel(R) Advanced Vector Extensions (Intel(R) AVX) instructions.
"calc_regression" assumes excess returns to calculate Information and Treynor Ratios
Intel MKL WARNING: Support of Intel(R) Streaming SIMD Extensions 4.2 (Intel(R) SSE4.2) enabled only processors has been deprecated. Intel oneAPI Math Kernel Library 2025.0 will require Intel(R) Advanced Vector Extensions (Intel(R) AVX) instructi

In [596]:
fx_pred.T

USUK      USEU      USSZ      USJP
Annualized Alpha                -0.007599  0.008869  0.049590 -0.001145
Interest Rate Differential Beta  0.110666 -1.633806 -2.066667  0.105380
R-Squared                        0.000020  0.004398  0.006156  0.000040

# 3.3 b)
    a) Sterling and Yen have positive Betas which means they will depreciate in value when their risk-free rates increase against USD rate. Thus, USD will strengthen against Sterling and Yen.
    
    b) Euro and Swiss Franc have negative Betas which means their value will appreciate when their risk-free rate strengthens against USD risk-free and they will be stronger than USD.
    
    c) When FX predictions are compared against each other, the one with the highest r-squared is USD vs Swiss Franc (0.006156). Although the r-squared is very low to base a meaning, relatively it's high against other currencies

# 3.4 a) - Predict Expected Excess Log Return 

In [658]:
fx_premium_list = []

for fx, rf in zip(fx_pred.T.columns, rf_rates_log.columns):
    temp_df = pd.DataFrame(columns = [f'Expect Excess Ret {fx[2:]}', f'Alpha {fx[2:]}', 'Constant', f'Beta {fx[2:]}', f'Interest Rate Diff {fx[2:]}'], index = rf_rates_log.index)

    temp_df[f'Alpha {fx[2:]}'] = fx_pred.loc[fx]['Annualized Alpha']/12
    temp_df[f'Beta {fx[2:]}'] = fx_pred.loc[fx]['Interest Rate Differential Beta']
    temp_df['Constant'] = -1
    temp_df[f'Interest Rate Diff {fx[2:]}'] = rf_usd_log - rf_rates_log[rf]
    temp_df[f'Expect Excess Ret {fx[2:]}'] = temp_df[f'Alpha {fx[2:]}'] + (temp_df[f'Beta {fx[2:]}'] + temp_df['Constant']) * temp_df[f'Interest Rate Diff {fx[2:]}'] 

    positive_ret = len(temp_df[temp_df[f'Expect Excess Ret {fx[2:]}']>0])
    total_months = len(temp_df)
    temp_index = [str(fx[2:])]
    
    temp_agg_df = pd.DataFrame(data = [[positive_ret, total_months, 100*positive_ret/total_months]], columns = ['Months Positive Return', 'Total Months', 'Freq of Positive Return (%)'], index = temp_index)

    fx_premium_list.append(temp_agg_df)
    
pd.concat(fx_premium_list)

Months Positive Return  Total Months  Freq of Positive Return (%)
UK                      70           274                    25.547445
EU                     140           274                    51.094891
SZ                     173           274                    63.138686
JP                       1           274                     0.364964

# 3.4 b) 
    For Swiss Franc and Euro, more than 50% of the time, the expected excess return is positive. For Sterling it's 25% of the time and for Yen it's nearly 0% of the time positive and 100% of the time negative

# 3.4 c)
    Since Yen loses its strength against USD, one can borrow in Yen with Yen's interest rate and convert that into USD and buy Treasury Bond. Then, at the maturity, one can sell the Treasury and collect the interest of USD. Finally, one can convert USD into Yen, pay his debt in Yen with its interest and keep the difference as the profit
    